### ISSUE TITLE
Program does not open Minecraft 1.13 files

### ISSUE THREAD

**Issue by erich666** [erich666/Mineways#35]
Since Mineways grew from minutor's code base, I've recorded [this issue at their site](https://github.com/mrkite/minutor/issues/125). I did some initial investigation, and it looks fairly straightforward (though tedious coding, translating strings to their block numbers) to read the new file format.

However, Mineways certainly is based on a limit that Minecraft has held to, 256 block types (and up to 16 sub-block data: sub-types, orientation, etc.). A 257th block doesn't fit. I suspect I could bolt on some clunky code that says "block 254 used, go look up its location in a table for this slice and you'll find the real block value there." This might keep Mineways going. Or maybe I'll call it a day on this 7-year-old program, as much joy as it's provided me.

**Comment by JMY1000** [comment]
I'd be awfully sad if development discontinued! I'm curious how [jmc2obj](https://github.com/jmc2obj/j-mc-2-obj) and [Chunky](https://github.com/llbit/chunky) plan to handle any changes; as far as I know, they're really the only big third party apps that really support an external renderer. Perhaps Minecraft will expand their official export capabilities, but for now, it really looks to be a "Microsoft made us" thing.

How difficult would a refactor to just use an arbitrary number of block types be?

**Comment by erich666** [comment]
> How difficult would a refactor to just use an arbitrary number of block types be?

Given that all data storage in my code is based on block IDs being bytes, a considerable bit. I could just make all block types shorts, but Mineways would then use like 1.5X more memory (the block data field would still be a byte). Hmmm, that's a thought, I might try some trickiness of using the block "data" field, which Minecraft uses only 4 bits of, to create extended block types. Sounds like a pain, either way. Anyway, short answer is that something non-trivial will need to be added to Mineways to allow this larger number of blocks. When I get to it I might simply say "tough luck, block IDs are now shorts, take the memory hit." This will be particularly unpleasant on the Mac, where only the 32 bits version works.

**Comment by JMY1000** [comment]
Is there anything that makes it problematic to just allocate 9 bits and expand in the future as necessary? Alternatively, does Mineways really need to load the entire selection into memory, or could a smaller swap be used? I've certainly run into the 32 bit limit on macOS and having lower memory usage as a whole would be nice.

**Comment by erich666** [comment]
> Is there anything that makes it problematic to just allocate 9 bits and expand in the future as necessary

Computers deal with multiples of 8 bits - 9 bits is a lot of crazy coding (which is what Minecraft is in fact aiming to do) and less efficient.

> Alternatively, does Mineways really need to load the entire selection into memory, or could a smaller swap be used?

For simplicity, yes. Mineways is for 3D printing, so does a lot of global consistency checks and fixes to the data. Not needed as much for rendering, so a chunk by chunk approach could be used (which is what I believe jmc2obj does), with the side effect of creating lots of little meshes, which can slow down the frame display rate inside modelers. That said, there's already a solution for the "out of memory" problem, though a little more work: [use scripting](http://www.realtimerendering.com/erich/minecraft/public/mineways/scripting.html#uie), which allows you to automatically perform a precise set of exports in a sequence. You then import each OBJ file produced.

Basically, it's software, anything's possible, but it's a hobby project. I think it's fun adding new blocks or functionality. I don't enjoy rewriting how data is stored just to get the system functioning like it was before Mojang/MS decided to change (and not bother documenting) their file format.

**Comment by JMY1000** [comment]
Maybe I'm missing something, but doesn't C support [allocating bit by bit?](https://stackoverflow.com/questions/284519/can-i-allocate-a-specific-number-of-bits-in-c)

**Comment by erich666** [comment]
Note that "divide by 8" in the page you point at. That's because malloc() is called using the number of bytes you want to allocate. So "(<number of data points>/8)+1" says "I want, say, 245 bits, so how many bytes do I need?" 245/8 + 1 = 31 bytes. (And, FWIW, that formula is inefficient. The right one is ((bits-1)/8+1).)

Anyway, like I say, adding a bit is possible, but things that were easy, e.g., "is the block ID[loc] == 112?" suddenly become more involved, "id = retrieveBlockIDFromArray(block,loc); is id == 112?" Unpleasant, and pervasive - the block ID is used all over the place. It also makes debugging a bit nightmarish, as you can't just look at the array of bytes and see what's in there, you have to always call the helper function to actually know the ID for a given location.

**Comment by erich666** [comment]
Resolved - quite an involved puzzle to figure out the new format.

**Comment by jssignore1** [comment]
Thanks for fixing this!
I'd been using Minutor years ago and just updated to 1.13, and came here from this issue you raised on their github. Just tried out Mineways and it works great with 1.13 :) Thanks, will be using this for sure now!
